src/oidc_child/oidc_child_curl.c

```diff
@@ -0,0 +1,450 @@
+/*
+    SSSD
+
+    Helper child for OIDC and OAuth 2.0 Device Authorization Grant
+    Curl based HTTP access
+
+    Authors:
+        Sumit Bose <sbose@redhat.com>
+
+    Copyright (C) 2022 Red Hat
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <unistd.h>
+
+#include <curl/curl.h>
+#include "util/util.h"
+#include "oidc_child/oidc_child_util.h"
+
+/* The curl write_callback will always append the received data. To start a
+ * new string call clean_http_data() before the curl request.*/
+void clean_http_data(struct devicecode_ctx *dc_ctx)
+{
+    talloc_free(dc_ctx->http_data);
+    dc_ctx->http_data = NULL;
+}
+
+static size_t write_callback(char *ptr, size_t size, size_t nmemb,
+                             void *userdata)
+{
+    size_t realsize = size * nmemb;
+    struct devicecode_ctx *data = (struct devicecode_ctx *) userdata;
+
+    DEBUG(SSSDBG_TRACE_ALL, "%*s\n", (int) realsize, ptr);
+
+    if (data->http_data == NULL) {
+        data->http_data = talloc_strndup(data, ptr, realsize);
+        if (data->http_data != NULL) {
+            talloc_set_destructor((void *) data->http_data,
+                                  sss_erase_talloc_mem_securely);
+        }
+    } else {
+        data->http_data = talloc_strndup_append(data->http_data, ptr, realsize);
+    }
+    if (data->http_data == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to copy received data.\n");
+        return 0;
+    }
+
+    return realsize;
+}
+
+static int libcurl_debug_callback(CURL *curl_ctx, curl_infotype type,
+                                  char *data, size_t size, void *userptr)
+{
+    static const char prefix[CURLINFO_END][3] = {
+                                     "* ", "< ", "> ", "{ ", "} ", "{ ", "} " };
+    char *str = NULL;
+
+    switch (type) {
+    case CURLINFO_TEXT:
+    case CURLINFO_HEADER_IN:
+    case CURLINFO_HEADER_OUT:
+        str = talloc_asprintf(NULL, "%s%.*s", prefix[type], (int) size, data);
+        if (str != NULL) {
+            sss_debug_fn(__FILE__, __LINE__, __FUNCTION__, SSSDBG_TRACE_ALL,
+                         "libcurl: %s", str);
```

Comment:
Would it be possible to avoid overhead of 'asprintf' (alloc + copy) and to print "libcurl: %s%.*s" directly?

Reply:
fixed